idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/frontend/api/fir/components/KtFirCallResolver.kt

```diff
@@ -111,23 +115,58 @@ internal class KtFirCallResolver(
-            is FirErrorReferenceWithCandidate -> calleeReference.createErrorCallTarget(source)
+        val target = calleeReference.createCallTarget() ?: return null
+        return KtFunctionCall(createArgumentMapping(), target)
+    }
+
+    private fun FirAnnotationCall.asAnnotationCall(): KtAnnotationCall? {
+        val target = calleeReference.createCallTarget() ?: return null
+        return KtAnnotationCall(createArgumentMapping(), target)
+    }
+
+    private fun FirDelegatedConstructorCall.asDelegatedConstructorCall(): KtDelegatedConstructorCall? {
+        val target = calleeReference.createCallTarget() ?: return null
+        return KtDelegatedConstructorCall(createArgumentMapping(), target, isThis)
+    }
+
+    private fun FirReference.createCallTarget(): KtCallTarget? {
+        return when (this) {
+            is FirResolvedNamedReference -> getKtFunctionOrConstructorSymbol()?.let { KtSuccessCallTarget(it) }
+            is FirErrorNamedReference -> createErrorCallTarget(source)
+            is FirErrorReferenceWithCandidate -> createErrorCallTarget(source)
             is FirSimpleNamedReference ->
                 null
-              /*  error(
-                    """
-                      Looks like ${this::class.simpleName} && it calle reference ${calleeReference::class.simpleName} were not resolved to BODY_RESOLVE phase,
-                      consider resolving it containing declaration before starting resolve calls
-                      ${this.render()}
-                      ${(this.psi as? KtElement)?.getElementTextInContext()}
-                      """.trimIndent()
-                )*/
-            else -> error("Unexpected call reference ${calleeReference::class.simpleName}")
-        } ?: return null
-        return KtFunctionCall(target)
+            /*  error(
+                  """
+                    Looks like ${this::class.simpleName} && it calle reference ${calleeReference::class.simpleName} were not resolved to BODY_RESOLVE phase,
+                    consider resolving it containing declaration before starting resolve calls
+                    ${this.render()}
+                    ${(this.psi as? KtElement)?.getElementTextInContext()}
+                    """.trimIndent()
+              )*/
+            else -> error("Unexpected call reference ${this::class.simpleName}")
+        }
+    }
+
+    private fun FirCall.createArgumentMapping(): LinkedHashMap<KtValueArgument, KtValueParameterSymbol> {
+        val ktArgumentMapping = LinkedHashMap<KtValueArgument, KtValueParameterSymbol>()
+        argumentMapping?.let {
+            for ((firExpression, firValueParameter) in it.entries) {
+                val parameterSymbol = firValueParameter.buildSymbol(firSymbolBuilder) as KtValueParameterSymbol
+                // Note: We use getNonStrictParentOfType() below because for spread and named arguments, the source is the KtValueArgument.
```

Comment:
Maybe let's introduce this logic in the code instead of looking for the nearest parent? Otherwise, there may be cases when we go up the PSI tree without finding the argument and we may not know about it. E.g, if the source elements  mapping in the fir tree will be changed

Reply:
Can you explain more what you mean by "introducing logic in the code"?

Reply:
I mean that it is better to explicitly check which kind of argument you have (spread, named, or other) and depending on the argument kind find it's source

Reply:
I updated the code, I'm not sure if I got what you were suggesting correctly. I personally think it's a little bit more unstable than just always looking up to find the parent (or some ancestor) `KtValueArgument`. Let me know if it looks good.